File: data_structures/linked_list/typescript/SinglyLinkedList.ts
import LLNode from '@ds/linked_list/typescript/SinglyLinkedListNode';
import { ILinkedList, ILinkedListNode, Nullable, Undefined } from './types';

export default class SinglyLinkedList<T> implements ILinkedList<T> {
    public length = 0;
    public head: Nullable<LLNode<T>>;
    public tail: Nullable<LLNode<T>>;

    constructor() {
        this.head = null;
        this.tail = null;
    }

    get isEmpty(): boolean {
        return !this.head;
    }

    toString(): string {
        return this.toArray().toString();
    }

    toArray(): T[] {
        if (this.isEmpty) {
            return [];
        }

        return this.head.toArray();
    }

    fromArray(array: T[]): this {
        const nodes = new LLNode<T>().fromArray(array);

        let tail = nodes;

        while (tail.next) {
            tail = tail.next;
        }

        this.head = nodes;
        this.tail = tail;
        this.length = array.length;

        return this;
    }

    /**
     * Removes an item from the end of Linked List
     */
    pop(): Undefined<T> {
        /**
         * If there is nothing on the list return undefined
         * Loop through until you get to the tail
         * Set the next property of the 2nd to last node to be null
         * Set the tail to be the 2nd to last node
         * Decrement the length
         * Return the value of the node removed
         */
        if (this.isEmpty) {
            return undefined;
        }

        let current = this.head;
        let newTail = this.tail;

        while (current.next) {
            newTail = current;
            current = current.next;
        }

        this.tail = newTail;

        if (this.tail) {
            this.tail.next = null;
        }

        this.length--;

        if (this.length == 0) {
            this.head = null;
            this.tail = null;
        }

        return current.val;
    }

    /**
     * Appends new Node to the beginning of the list
     * @param value Value to be set on a node
     * @param key Key to be set on a node
     */
    unshift(value: T, key?: number | string): this {
        const node = new LLNode<T>(value, key);

        if (this.isEmpty) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head = node;
        }

        this.length++;
        return this;
    }

    /**
     * Removes Node from beginning
     * Returns removed Node or undefined
     */
    shift(): T {
        if (this.isEmpty) {
            return undefined;
        }

        const currentHead = this.head;
        this.head = currentHead.next;

        this.length--;

        if (this.length === 0) {
            this.head = null;
            this.tail = null;
        }

        return currentHead.val;
    }

    /**
     * Creates new Node using passed value
     * @param value Value to be set on a node
     * @param key Key to be set on a node
     */
    push(value: T, key?: number | string): this {
        /**
         * Create a new Node
         * Edge case: If there is no Head or Tail
         *  set then set Head and Tail to be newly created node
         *  Otherwise set the next property on the tail
         *  and tail property to be newly created node
         */
        const node = new LLNode<T>(value, key);

        if (this.isEmpty) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }

        this.length++;
        return this;
    }

    find(key: number | string): ILinkedListNode<T> {
        if (this.isEmpty) {
            return null;
        }

        let current = this.head;

        while (current.key != key) {
            if (current.next) {
                current = current.next;
            } else {
                return null;
            }
        }

        return current;
    }
}

File: data_structures/linked_list/python/singly_linked_list.py
from typing import Union, TypeVar, Generic, List
from data_structures.linked_list.python.singly_linked_list_node import ListNode


T = TypeVar("T")
ListNodeType = Union[ListNode[T], None]


class SinglyLinkedList(Generic[T]):
    length: int
    head: ListNodeType
    tail: ListNodeType

    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    @property
    def is_empty(self) -> bool:
        if self.length == 0:
            return True

        return False

    def from_array(self, nodes: List[T]) -> 'SinglyLinkedList':
        node = ListNode()
        node.from_array(nodes)

        tail = node

        while tail.next:
            tail = tail.next

        self.length = len(nodes)
        self.head = node
        self.tail = tail

        return self

    def to_array(self, to_string=False) -> List[T]:
        if self.is_empty:
            return []

        return self.head.to_array(to_string)

    def pop(self) -> T:
        if self.is_empty:
            return None

        current = self.head
        new_tail = None

        while current.next:
            new_tail = current
            current = current.next

        self.length -= 1
        self.tail = new_tail

        if self.tail:
            self.tail.next = None

        if self.is_empty:
            self.head = None
            self.tail = None

        return current.val

    def unshift(self, value: T, key: int = None) -> 'SinglyLinkedList':
        node = ListNode(val=value, key=key)

        if self.is_empty:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head = node

        self.length += 1

        return self

    def shift(self) -> T:
        if self.is_empty:
            return None

        current_head = self.head
        self.head = current_head.next

        self.length -= 1

        if self.length == 0:
            self.head = None
            self.tail = None

        return current_head.val

    def push(self, val: T, key: int = None) -> 'SinglyLinkedList':
        node = ListNode(key=key, val=val)

        if self.is_empty:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self.length += 1
        return self

    def find(self, key: int) -> Union[T, None]:
        if self.is_empty:
            return None

        current = self.head

        while current.key != key:
            if current.next:
                current = current.next
            else:
                return None

        return current

    def __repr__(self):
        result = self.to_array(to_string=True)
        result = " -> ".join(result)
        return f"LinkedList({result})"
